File: src/ballbot_controller/src/simple_controller.cpp

// BUG FIX 1: Removed hardcoded absolute include path.
// Use the package-relative include path that CMake/colcon sets up.
#include "ballbot_controller/simple_controller.hpp"
 
#include "geometry_msgs/msg/twist_stamped.hpp"
#include "std_msgs/msg/float64_multi_array.hpp"
#include <eigen3/Eigen/Geometry>
#include <eigen3/Eigen/src/Core/Matrix.h>
#include <rclcpp/logging.hpp>
 
Simplecontroller::Simplecontroller(const std::string &name) : Node(name) {
 
  // BUG FIX 2: Parameter names must match what controller_launch.py passes:
  //   "wheel_radius"    → already correct
  //   "wheel_base"      → WRONG; launch sends "wheel_separation", not "wheel_base"
  // Default values also corrected to match URDF/YAML (radius=0.17, sep=0.33).
  wheel_radius_ = declare_parameter("wheel_radius", 0.17);
  wheel_base_   = declare_parameter("wheel_separation", 0.33);  // was "wheel_base"
 
  RCLCPP_INFO_STREAM(get_logger(), "Using wheel radius "     << wheel_radius_);
  RCLCPP_INFO_STREAM(get_logger(), "Using wheel separation " << wheel_base_);
 
  // BUG FIX 3: Subscriber topic was "/ballbot/cmd_vel" (non-existent).
  // The correct topic is "/ballbot_controller/cmd_vel" — this is what
  // twist_relay publishes (TwistStamped) for the simple_controller path.
  sub_ = this->create_subscription<geometry_msgs::msg::TwistStamped>(
      "/ballbot_controller/cmd_vel", 10,
      std::bind(&Simplecontroller::callback, this, std::placeholders::_1));
 
  pub_ = this->create_publisher<std_msgs::msg::Float64MultiArray>(
      "/simple_velocity_controller/commands", 10);
 
  // Kinematic conversion matrix  [v_left, v_right]^T = M^{-1} * [v_lin, omega]^T
  pose_ << wheel_radius_ / 2,  wheel_radius_ / 2,
           wheel_radius_ / wheel_base_, -wheel_radius_ / wheel_base_;
 
  RCLCPP_INFO_STREAM(get_logger(), "Conversion matrix:\n" << pose_);
}
 
void Simplecontroller::callback(
    const geometry_msgs::msg::TwistStamped::SharedPtr msg) {
  std_msgs::msg::Float64MultiArray wheel_speed_msg;
 
  Eigen::Vector2d vel(msg->twist.linear.x, msg->twist.angular.z);
  Eigen::Vector2d wheel_vel = pose_.inverse() * vel;
 
  wheel_speed_msg.data.push_back(wheel_vel[0]);
  wheel_speed_msg.data.push_back(wheel_vel[1]);
 
  pub_->publish(wheel_speed_msg);
}
 
int main(int argc, char *argv[]) {
  rclcpp::init(argc, argv);
  auto node = std::make_shared<Simplecontroller>("simple_controller");
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}
 

File: src/ballbot_controller/ballbot_controller/simple_controller.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from std_msgs.msg import Float64MultiArray
from geometry_msgs.msg import TwistStamped
import numpy as np


class SimpleController(Node):

    def __init__(self):
        super().__init__("simple_controller")

        # BUG FIX 1: Default values were wrong AND swapped
        # Old: wheel_radius=0.033, wheel_separation=0.17  ← completely wrong
        # Correct values must match URDF and ballbot_controllers.yaml
        self.declare_parameter("wheel_radius", 0.17)
        self.declare_parameter("wheel_separation", 0.33)

        self.wheel_radius_ = self.get_parameter("wheel_radius").get_parameter_value().double_value
        self.wheel_separation_ = self.get_parameter("wheel_separation").get_parameter_value().double_value

        # BUG FIX 2: %d is integer format — prints 0 for floats always
        # Use %f or f-strings instead
        self.get_logger().info(f"Using wheel radius {self.wheel_radius_}")
        self.get_logger().info(f"Using wheel separation {self.wheel_separation_}")

        # BUG FIX 3: Missing leading slash — topics were relative to node namespace
        # and might not connect to the correct absolute topics
        self.wheel_cmd_pub_ = self.create_publisher(
            Float64MultiArray, "/simple_velocity_controller/commands", 10)

        self.vel_sub_ = self.create_subscription(
            TwistStamped, "/ballbot_controller/cmd_vel", self.velCallback, 10)

        # Kinematic conversion matrix
        # [v_lin, omega]^T = M * [v_left, v_right]^T
        # Invert to get wheel speeds from robot speeds
        self.speed_conversion_ = np.array([
            [self.wheel_radius_ / 2,              self.wheel_radius_ / 2],
            [self.wheel_radius_ / self.wheel_separation_, -self.wheel_radius_ / self.wheel_separation_]
        ])
        self.get_logger().info(f"The conversion matrix is:\n{self.speed_conversion_}")

    def velCallback(self, msg: TwistStamped):
        robot_speed = np.array([
            [msg.twist.linear.x],
            [msg.twist.angular.z]
        ])
        wheel_speed = np.linalg.inv(self.speed_conversion_) @ robot_speed

        wheel_speed_msg = Float64MultiArray()

        # BUG FIX 4: Wheel order was swapped [1,0] instead of [0,1]
        # simple_velocity_controller expects [left, right]
        # matching the joints order in ballbot_controllers.yaml:
        #   joints: [wheel_left_joint, wheel_right_joint]
        wheel_speed_msg.data = [wheel_speed[0, 0], wheel_speed[1, 0]]

        self.wheel_cmd_pub_.publish(wheel_speed_msg)


def main():
    rclpy.init()
    simple_controller = SimpleController()
    rclpy.spin(simple_controller)
    simple_controller.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()
